javascript/ql/src/Security/CWE-020/IncompleteUrlRegExp.ql

```diff
@@ -0,0 +1,68 @@
+/**
+ * @name Incomplete URL regular expression
+ * @description Using a regular expression that contains an 'any character' may match more URLs than expected.
+ * @kind problem
+ * @problem.severity error
+ * @precision high
+ * @id js/incomplete-url-regexp
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-20
+ */
+
+import javascript
+private import semmle.javascript.RegularExpressions
+
+module IncompleteUrlRegExpTracking {
+
+  /**
+   * A taint tracking configuration for incomplete URL regular expressions sources.
+   */
+  class Configuration extends TaintTracking::Configuration {
+    Configuration() { this = "IncompleteUrlRegExpTracking" }
+
+    override
+    predicate isSource(DataFlow::Node source) {
+      isIncompleteHostNameRegExpPattern(source.asExpr().getStringValue(), _)
+    }
+
+    override
+    predicate isSink(DataFlow::Node sink) {
+      isMaybeARegularExpressionSource(sink)
+    }
+
+  }
+
+}
+
+/**
+ * Holds if `pattern` is a regular expression pattern for URLs with a host matched by `hostPart`,
+ * and `pattern` contains a subtle mistake that allows it to match unexpected hosts.
+ */
+bindingset[pattern]
+predicate isIncompleteHostNameRegExpPattern(string pattern, string hostPart) {
+  hostPart = pattern.regexpCapture(
+    "(?i).*" +
+    // an unescaped single `.`
+    "(?<!\\\\)[.]" +
+    // immediately followed by a sequence of subdomains, perhaps with some regex characters mixed in, followed by a known TLD
+    "([():|?a-z0-9-]+(\\\\)?[.](com|org|edu|gov|uk|net))" +
+    ".*", 1)
```

Comment:
Couldn't you incorporate the condition from line 65 here as a negative lookahead?

Reply:
Line 65 is used for the pragmatic whitelisting of practically benign TPs, I would prefer to keep that out of the principled part of the query.